### PR TITLE
Add `from_arrow_host` functions for cudf interop with nanoarrow

### PR DESCRIPTION
**Pull request by zeroshade**
## Description
Following up from #15458 and continuing the work to address #14926 adding host memory version of `from_arrow_device` which will perform the copies from host memory to create cudf objects.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by zeroshade** [comment]
@vyasr @davidwendt I've got this up! There's still another test or two I need to add, but I figured it would be good to get this PR filed before the weekend and get some eyes on it.

File: cpp/src/interop/from_arrow_device.cu [hunk 1]
```diff
@@ -88,6 +93,309 @@ data_type arrow_to_cudf_type(const ArrowSchemaView* arrow_view)
...
+
+  return input->offset == 0
+           ? std::move(out_col)
+           : std::make_unique<column>(
+               cudf::detail::slice(out_col->view(),
+                                   static_cast<size_type>(input->offset),
+                                   static_cast<size_type>(input->offset + input->length),
+                                   stream),
+               stream,
+               mr);
```
Comment (davidwendt) [inline cpp/src/interop/from_arrow_device.cu:267]:
```suggestion
  auto bitmask =  array->buffers[validity_buffer_idx]; // not sure this is right
  auto out_col  = cudf::column_view(
                        data_type{type_id::STRING}
                        num_rows,
                        chars_column->view().head(),
                        static_cast<bitmask_type*>(bitmask),
                        input->null_count,
                        {offsets_column->view()});

  return std::make_unique<column>(
               cudf::detail::slice(out_col,
                                   static_cast<size_type>(input->offset),
                                   static_cast<size_type>(input->offset + input->length),
                                   stream),
               stream,
               mr);
```
Create a view to prevent making two columns when `offset!=0`
I think that something similar could be done with the offsets column since we only need the view to build the strings' column view.

File: cpp/src/interop/from_arrow_device.cu [hunk 2]
```diff
@@ -88,6 +93,309 @@ data_type arrow_to_cudf_type(const ArrowSchemaView* arrow_view)
...
+
+  return num_rows == input->length
+           ? std::move(out_col)
+           : std::make_unique<column>(
+               cudf::detail::slice(out_col->view(),
+                                   static_cast<size_type>(input->offset),
+                                   static_cast<size_type>(input->offset + input->length),
+                                   stream),
+               stream,
+               mr);
```
Comment (davidwendt) [inline cpp/src/interop/from_arrow_device.cu:396]: Same comment as above about using views instead of multiple columns

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/src/interop/from_arrow_device.cu [hunk 3]
```diff
@@ -88,6 +93,310 @@ data_type arrow_to_cudf_type(const ArrowSchemaView* arrow_view)
...
+  ArrowSchemaView view;
+  NANOARROW_THROW_NOT_OK(ArrowSchemaViewInit(&view, offset_schema.get(), nullptr));
+  auto offsets_column =
+    this->operator()<int32_t>(&view, &offsets_array, data_type(type_id::INT32), true);
+  NANOARROW_THROW_NOT_OK(ArrowSchemaViewInit(&view, char_data_schema.get(), nullptr));
+  auto chars_column = this->operator()<int8_t>(&view, &char_array, data_type(type_id::INT8), true);
+
+  auto const num_rows = offsets_column->size() - 1;
+  auto bitmask        = get_mask_buffer(input);
+  auto out_col        = cudf::column_view(data_type{type_id::STRING},
```
Comment (davidwendt) [inline cpp/src/interop/from_arrow_device.cu:254]: Sorry about this but I think your original implementation was better than my suggestion here.
I was thinking the data was already copied to the device and we could just wrap it with a `column_view` but the copy of the offsets, bitmask, etc need to occur and now those are double-copied because of my suggestion.
The original implementation only did the double-copy if the offset!=0 and so I'd like to return to that implementation.
Sorry again.

Comment (zeroshade) [reply cpp/src/interop/from_arrow_device.cu]: updated back to the original implementation

File: cpp/src/interop/from_arrow_device.cu [hunk 4]
```diff
@@ -88,6 +93,309 @@ data_type arrow_to_cudf_type(const ArrowSchemaView* arrow_view)
 
 namespace {
 
+struct dispatch_copy_from_arrow_device {
```
Comment (davidwendt) [inline cpp/src/interop/from_arrow_device.cu:96]: Could the new code be separated into its own source file?
Is there some common code here that would need to be refactored?
```suggestion
struct dispatch_copy_from_arrow_host {
```

**Review by davidwendt — CHANGES_REQUESTED** [review]
Perhaps rename the `from_arrow_device_host_test.cpp` to `from_arrow_host_test.cpp`?

File: cpp/src/interop/from_arrow_device.cu [hunk 5]
```diff
@@ -39,6 +42,8 @@
 #include <nanoarrow/nanoarrow.h>
 #include <nanoarrow/nanoarrow.hpp>
 
+#include <iostream>
```
Comment (pmattione-nvidia) [inline cpp/src/interop/from_arrow_device.cu:45]: Is this header still needed?

File: cpp/tests/interop/from_arrow_device_host_test.cpp [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+#include <cudf/dictionary/dictionary_column_view.hpp>
+#include <cudf/dictionary/dictionary_factories.hpp>
+#include <cudf/dictionary/encode.hpp>
+#include <cudf/interop.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
```
Comment (pmattione-nvidia) [inline cpp/tests/interop/from_arrow_device_host_test.cpp:42]: This whole file could use a lot more comments explaining what exactly is happening and getting tested; there are very few comments here.

File: cpp/src/interop/from_arrow_device.cu (hunk 4 above)
Comment (zeroshade) [reply cpp/src/interop/from_arrow_device.cu]: file renamed, struct renamed, and new code separated to its own source file. unfortunately there's not much in the way of common code that can be easily refactored between them without quite a lot of boilerplate, `if constexpr` conditional stuff, etc.

File: cpp/tests/interop/from_arrow_device_host_test.cpp (hunk 6 above)
Comment (zeroshade) [reply cpp/tests/interop/from_arrow_device_host_test.cpp]: I've added some more comments to this. if there are specific areas where it's unclear what is going on, let me know and i'll add some more comments to those areas. I had figured the test names and code was sufficient, but that might be because i'm already familiar with a lot of the nano arrow code patterns

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/tests/interop/nanoarrow_utils.hpp [hunk 7]
```diff
@@ -135,7 +137,193 @@ void populate_dict_from_col(ArrowArray* arr, cudf::dictionary_column_view dview)
...
+    tmp->null_count =
+      data.size() -
+      ArrowBitCountSet(ArrowArrayValidityBitmap(tmp.get())->buffer.data, 0, mask.size());
+  }
+
+  ArrowBuffer buf;
+  ArrowBufferInit(&buf);
+  NANOARROW_THROW_NOT_OK(
+    ArrowBufferAppend(&buf, reinterpret_cast<void const*>(data.data()), sizeof(T) * data.size()));
+  ArrowArraySetBuffer(tmp.get(), 1, &buf);
```
Comment (davidwendt) [inline cpp/tests/interop/nanoarrow_utils.hpp:195]: Calls to `ArrowArraySetBuffer` must be wrapped with `NANOARROW_THROW_NOT_OK` or something similar that checks the return code otherwise this will fail in a Debug build.

File: cpp/src/interop/from_arrow_host.cu [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+{
+  CUDF_EXPECTS(schema != nullptr && input != nullptr,
+               "input ArrowSchema and ArrowDeviceArray must not be NULL");
+  CUDF_EXPECTS(input->device_type == ARROW_DEVICE_CPU,
+               "ArrowDeviceArray must have CPU device type for `from_arrow_host`");
+
+  CUDF_FUNC_RANGE();
+
+  ArrowSchemaView view;
+  NANOARROW_THROW_NOT_OK(ArrowSchemaViewInit(&view, schema, nullptr));
```
[the remark this answers is not in the document]
Comment (zeroshade) [reply cpp/src/interop/from_arrow_host.cu:422]: sure. Any particular reason?

[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/interop/from_arrow_host.cu]: This is our normal pattern. https://github.com/rapidsai/cudf/blob/branch-24.06/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#libcudf-api-and-implementation
The public API does the nvtx range and calls a detail function with a similar signature with no default parameters.
The `detail` functions may also be called by other internal functions and we would want those paths to also check parameters, etc.

File: cpp/src/interop/from_arrow_device.cu [hunk 9]
```diff
@@ -16,14 +16,17 @@
...
 #include <cudf/column/column_view.hpp>
 #include <cudf/copying.hpp>
+#include <cudf/detail/copy.hpp>
 #include <cudf/detail/get_value.cuh>
 #include <cudf/detail/interop.hpp>
 #include <cudf/detail/null_mask.hpp>
 #include <cudf/detail/nvtx/ranges.hpp>
 #include <cudf/detail/transform.hpp>
 #include <cudf/detail/unary.hpp>
+#include <cudf/dictionary/dictionary_factories.hpp>
```
Comment (davidwendt) [inline cpp/src/interop/from_arrow_device.cu:29]: I believe these new includes can be removed now.

File: cpp/tests/interop/from_arrow_host_test.cpp [hunk 10]
```diff
@@ -0,0 +1,767 @@
...
+#include <cudf_test/testing_main.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/interop.hpp>
+#include <cudf/detail/iterator.cuh>
```
Comment (davidwendt) [inline cpp/tests/interop/from_arrow_host_test.cpp:32]:
```suggestion
```
I'm not sure these are needed.

File: cpp/src/interop/from_arrow_host.cu [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+
+  nanoarrow::UniqueSchema char_data_schema;
+  NANOARROW_THROW_NOT_OK(ArrowSchemaInitFromType(char_data_schema.get(), NANOARROW_TYPE_INT8));
+
+  ArrowSchemaView view;
+  NANOARROW_THROW_NOT_OK(ArrowSchemaViewInit(&view, offset_schema.get(), nullptr));
+  auto offsets_column =
+    this->operator()<int32_t>(&view, &offsets_array, data_type(type_id::INT32), true);
+  NANOARROW_THROW_NOT_OK(ArrowSchemaViewInit(&view, char_data_schema.get(), nullptr));
+  auto chars_column = this->operator()<int8_t>(&view, &char_array, data_type(type_id::INT8), true);
```
Comment (vyasr) [inline cpp/src/interop/from_arrow_host.cu:204]: IIUC the reason that the first elements of the length-2 buffer arrays inserted as the buffers to `offsets_buffers` and `char_buffers` above are set to nullptr is because those are the mask and we're relying on the mask to get set by these calls to dispatch to the int32/int8 overloads of this operator. Is that right? If so, could we add a comment explaining how this overload is leveraging overloads for other data types? I had to read this twice before I realized what was going on.

File: cpp/src/interop/from_arrow_host.cu [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+                                std::move(indices_column),
+                                std::move(*(column_contents.null_mask)),
+                                input->null_count);
+}
+
+template <>
+std::unique_ptr<column> dispatch_copy_from_arrow_host::operator()<cudf::struct_view>(
+  ArrowSchemaView* schema, ArrowArray const* input, data_type type, bool skip_mask)
+{
+  char buffer[1024];
```
Comment (vyasr) [inline cpp/src/interop/from_arrow_host.cu:273]: Is this used anywhere? I see it being passed to the lambda, but it's not accessed there.

File: cpp/src/interop/from_arrow_host.cu [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+                   auto type = arrow_to_cudf_type(&view);
+
+                   auto out = get_column_copy(&view, child, type, false, stream, mr);
+                   return std::make_unique<column>(
+                     cudf::detail::slice(out->view(),
+                                         static_cast<size_type>(input->offset),
+                                         static_cast<size_type>(input->offset + input->length),
+                                         stream),
+                     stream,
+                     mr);
```
Comment (vyasr) [inline cpp/src/interop/from_arrow_host.cu:292]: Can we eschew this call when the offset is 0?

File: cpp/src/interop/from_arrow_host.cu [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+    }
+
+    return col;
+  }
+};
+
+// forward declaration is needed because `type_dispatch` instantiates the
+// dispatch_copy_from_arrow_host struct causing a recursive situation for struct,
+// dictionary and list_view types.
+std::unique_ptr<column> get_column_copy(ArrowSchemaView* schema,
```
Comment (vyasr) [inline cpp/src/interop/from_arrow_host.cu:123]: Could we also have a brief comment on when this function is needed? It's not immediately obvious without inspecting all the call sites since it's basically a wrapper around the dispatch functor with some special handling for empty columns

**Review by vyasr — CHANGES_REQUESTED** [review]
Looks good overall. Do you also plan to include an overload that accepts an ArrowArray? I imagine it would be trivial but it would be handy for callers. That would be the most direct replacement for the existing `from_arrow` host API I think.

File: cpp/src/interop/from_arrow_host.cu (hunk 11 above)
Comment (zeroshade) [reply cpp/src/interop/from_arrow_host.cu:204]: added some comments. let me know if I should add more

File: cpp/src/interop/arrow_utilities.hpp [hunk 15]
```diff
@@ -26,5 +30,7 @@ namespace detail {
 static constexpr int validity_buffer_idx         = 0;
 static constexpr int fixed_width_data_buffer_idx = 1;
 
+data_type arrow_to_cudf_type(const ArrowSchemaView* arrow_view);
```
[the remark this answers is not in the document]
Comment (zeroshade) [reply cpp/src/interop/arrow_utilities.hpp:33]: sure, i'll create an `arrow_utilities.cpp` file to correspond with the `arrow_utilites.hpp` file

**Comment by zeroshade** [comment]
> Looks good overall. Do you also plan to include an overload that accepts an ArrowArray? I imagine it would be trivial but it would be handy for callers. That would be the most direct replacement for the existing from_arrow host API I think.

That's fair. I've added a `from_arrow` overload that takes an `ArrowSchema` and `ArrowArray`, do you think I should also add another overload for `from_arrow_column` too?

File: cpp/src/interop/from_arrow_host.cu [hunk 16]
```diff
@@ -0,0 +1,445 @@
...
+
+template <>
+std::unique_ptr<column> dispatch_copy_from_arrow_host::operator()<bool>(ArrowSchemaView* schema,
+                                                                        ArrowArray const* input,
+                                                                        data_type type,
+                                                                        bool skip_mask)
+{
+  auto data_buffer         = input->buffers[fixed_width_data_buffer_idx];
+  const auto buffer_length = bitmask_allocation_size_bytes(input->length + input->offset);
+  // mask-to-bools expects the mask to be bitmask_type aligned/padded
```
[the remark this answers is not in the document]
Comment (zeroshade) [reply cpp/src/interop/from_arrow_host.cu:144]: fair, i can remove this comment. its unnecessary

**Comment by davidwendt** [comment]
There is already a `to_arrow_utilities.cpp/.hpp`. 
Could we just move the function declared/defined there into this new `arrow_utilities.cpp/.hpp`? 
And then delete the `to_arrow_utilities.cpp/.hpp` files.

**Comment by vyasr** [comment]
> That's fair. I've added a `from_arrow` overload that takes an `ArrowSchema` and `ArrowArray`, do you think I should also add another overload for `from_arrow_column` too?

It's definitely not critical, but I also think we may as well do so while we're at it for completeness.

**Comment by zeroshade** [comment]
@vyasr Added `from_arrow_column` overload. Let me know if there's any other changes needed! :smile:

File: cpp/tests/interop/nanoarrow_utils.hpp [hunk 17]
```diff
@@ -135,7 +137,193 @@ void populate_dict_from_col(ArrowArray* arr, cudf::dictionary_column_view dview)
...
+
+  ArrowBuffer buf;
+  ArrowBufferInit(&buf);
+  NANOARROW_THROW_NOT_OK(
+    ArrowBufferAppend(&buf, reinterpret_cast<void const*>(data.data()), sizeof(T) * data.size()));
+  NANOARROW_THROW_NOT_OK(ArrowArraySetBuffer(tmp.get(), 1, &buf));
+
+  tmp->length = data.size();
+
+  return tmp;
```
Comment (vyasr) [inline cpp/tests/interop/nanoarrow_utils.hpp:199]: I'm not quite sure how this is working right now. UniqueArrays aren't copyable, are they? So how does returning this local object work here without an explicit `std::move`? Are we relying on copy-elision to happen here? That isn't guaranteed in this scenario where `tmp` is a locally declared variable; C++17 only guarantees it when returning a temporary. Am I missing something, or are we just getting lucky here because of some implementation-specific behavior of the chosen host compiler?

The same question applies to all of the SFINAE overloads of this function.

File: cpp/tests/interop/from_arrow_host_test.cpp [hunk 18]
```diff
@@ -0,0 +1,763 @@
...
+#include <cudf/interop.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+// create a cudf::table and equivalent arrow table with host memory
+std::tuple<std::unique_ptr<cudf::table>, nanoarrow::UniqueSchema, nanoarrow::UniqueArray>
+get_nanoarrow_host_tables(cudf::size_type length)
```
Comment (vyasr) [inline cpp/tests/interop/from_arrow_host_test.cpp:41]: Can this be easily shared with the version in `to_arrow_device_test.cpp`? It looks very nearly like a carbon copy, but I didn't check it quite line-for-line. There's extra logic for host-device transfer, but maybe that can be done as a post-processing step so that one function can call the other?

**Review by vyasr — COMMENTED** [review]
A few small nits, mostly around testing, then I think I'm happy with this!

**Comment by vyasr** [comment]
> @vyasr Added `from_arrow_column` overload. Let me know if there's any other changes needed! 😄

Sorry, I forgot to submit my review earlier! However, I did see that there was a failure on the struct test in [the last run](https://github.com/rapidsai/cudf/actions/runs/9197583117/job/25299393341), so there's some edge case that's off. Not sure if it's in the test or in the code though; from quick inspection I lean towards the test.

File: cpp/tests/interop/nanoarrow_utils.hpp (hunk 17 above)
Comment (davidwendt) [reply cpp/tests/interop/nanoarrow_utils.hpp:199]: We do rely on copy-elision like this in many places in libcudf. 
Since a unique-ptr object cannot be copied, the compiler has no choice but report an error if copy-elision is not possible.

Comment (vyasr) [reply cpp/tests/interop/nanoarrow_utils.hpp:199]: Fair enough, I guess it's OK to rely on the optional behavior behavior in this kind of scenario since there isn't anything that can go wrong other than a failure to compile, which we would see immediately. No risk of UB.

File: cpp/src/interop/from_arrow_device.cu [hunk 19]
```diff
@@ -379,11 +336,25 @@ dispatch_tuple_t get_column(ArrowSchemaView* schema,
 
 }  // namespace
 
-unique_table_view_t from_arrow_device(ArrowSchemaView* schema,
+unique_table_view_t from_arrow_device(ArrowSchema const* schema,
```
[the remark this answers is not in the document]
Comment (zeroshade) [reply cpp/src/interop/from_arrow_device.cu:339]: This was because of @davidwendt's request that we shift the view init into the function impl inside the `detail` for other callers. see https://github.com/rapidsai/cudf/pull/15645#discussion_r1602115746

[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/interop/from_arrow_device.cu:339]: Right. The view init was happening in the public API definition and this just moves it to the `detail` implementation instead I believe.

File: cpp/tests/interop/from_arrow_host_test.cpp (hunk 18 above)
Comment (zeroshade) [reply cpp/tests/interop/from_arrow_host_test.cpp:41]: It's not *that* easy to share since we generate the data for the columns and then use that for constructing both the nanoarrow and the cudf tables. I'll see if I can potentially wrap that in a helper that can be more easily shared

Comment (vyasr) [reply cpp/tests/interop/from_arrow_host_test.cpp:41]: Looks like you pulled it off to me, thanks!

**Review by vyasr — APPROVED** [review]

**Comment by vyasr** [comment]
Style again :(

**Review by davidwendt — APPROVED** [review]

**Comment by vyasr** [comment]
Yup, that's exactly what I'm suggesting. It would make maintaining easier, and it would also make testing easier since we'd effectively only have one real conversion test path to test (and it facilitates testing from higher level bindings in Java and Python where we can easily generate Arrow host data but not device data right now). But yeah we don't need to wait to merge this PR.

**Comment by zeroshade** [comment]
I don't think the test failures are my fault here. Is there anything else needed on my end to get this merged?

**Comment by vyasr** [comment]
No, nothing on your end. Unfortunately we're just having some CI outages. You're all good here.
